File: 2024/d06/src/main.py
input = open("src/input").read().strip()

# list of chars since cannot set index in string
grid = [list(line) for line in input.splitlines()]
rows = len(grid)
cols = len(grid[0])
dirs = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def find():
    for r in range(rows):
        for c in range(cols):
            x = grid[r][c]
            if x != "." and x != "#":
                return r, c
    raise Exception("?")


def part1():
    r, c = find()
    d = 0

    seen = set()
    while True:
        seen.add((r, c))
        dr, dc = dirs[d]
        nr, nc = r + dr, c + dc
        if not (0 <= nr < rows and 0 <= nc < cols):
            break
        if grid[nr][nc] == "#":
            d = (d + 1) % 4
        else:
            r, c = nr, nc
    return len(seen)


def part2():
    r, c = find()

    def is_loop(grid: list[list[str]], r: int, c: int):
        d = 0
        dr, dc = dirs[d]
        seen = set()
        while True:
            if (r, c, dr, dc) in seen:
                return True
            seen.add((r, c, dr, dc))
            nr, nc = r + dr, c + dc
            if not (0 <= nr < rows and 0 <= nc < cols):
                return False
            if grid[nr][nc] == "#":
                d = (d + 1) % 4
                dr, dc = dirs[d]
            else:
                r, c = nr, nc

    out = 0
    for i in range(rows):
        for j in range(cols):
            if grid[i][j] != ".":
                continue
            grid[i][j] = "#"
            if is_loop(grid, r, c):
                out += 1
            grid[i][j] = "."
    return out


from time import time

start = time()
print(f"{part2()=} took {time() - start}s")

File: 2024/d06/src/main.cpp
#include <bits/stdc++.h>
using namespace std;

// Contains. `c in a..b`
#define cts(a, b, c) (a <= c && c < b)

int main() {
  vector<string> grid;
  string line;
  while (cin >> line) grid.push_back(line);
  int rows = grid.size();
  int cols = grid[0].size();

  pair<int, int> me{-1, -1};
  iter(r, 0, rows) {
    iter(c, 0, cols) {
      if (grid[r][c] == '^') {
        me = {r, c};
        cout << "me " << r << ' ' << c << endl;
        break;
      }
    }
  }

  pair<int, int> dirs[4] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
  auto is_loop = [&](pair<int, int> me) {
    int d = 0;
    vector<bool> seen(rows * cols * 4);
    loop {
      int k = (me.first * cols + me.second) * 4 + d;
      if (seen[k]) return true;
      seen[k] = true;

      int nr = me.first + dirs[d].first;
      int nc = me.second + dirs[d].second;
      if (!(cts(0, rows, nr) && cts(0, cols, nc))) {
        return false;
      }
      if (grid[nr][nc] == '#') {
        d = (d + 1) % 4;
      } else {
        me = {nr, nc};
      }
    }
  };

  int count = 0;
  iter(r, 0, rows) {
    iter(c, 0, cols) {
      if (grid[r][c] == '.') {
        grid[r][c] = '#';
        if (is_loop(me)) count++;
        grid[r][c] = '.';
      }
    }
  }

  cout << count << endl;
}
